### PR TITLE
[new] ResizeSensor component

### PR DESCRIPTION
**Pull request by giladgray**
- `<ResizeSensor>` provides a React interface to `ResizeObserver` with a required `onResize` callback.
- refactor `OverflowList` and `Popover` to use this new guy.

### CODE REVIEW

File: packages/core/src/components/popover/popover.tsx [hunk 1]
```diff
@@ -350,7 +335,11 @@ export class Popover extends AbstractPureComponent<IPopoverProps, IPopoverState>
             disabled: isOpen && Utils.isElementOfType(rawTarget, Tooltip) ? true : rawTarget.props.disabled,
             tabIndex: this.props.openOnTargetFocus && isHoverInteractionKind ? tabIndex : undefined,
         });
-        return React.createElement(targetTagName, targetProps, clonedTarget);
+        return (
+            <ResizeSensor onResize={this.handlePopoverResize}>
```
Comment (giladgray) [inline packages/core/src/components/popover/popover.tsx]: this is new and awesome and so simple. measures the target too!

File: packages/core/test/tsconfig.json [hunk 2]
```diff
@@ -2,6 +2,14 @@
...
         "declaration": false,
-        "noEmit": true
+        "noEmit": true,
+        "lib": [
+          "dom",
+          "dom.iterable",
+          "es5",
+          "es2015.collection",
+          "es2015.iterable",
+          "es2015.promise"
```
Comment (giladgray) [inline packages/core/test/tsconfig.json]: added Promise types to tests to simplify the nested timeouts

File: packages/core/test/resize-sensor/resizeSensorTests.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        return (wrapper = mount<IResizeTesterProps>(
+            <ResizeTester onResize={onResize} />,
+            // must be in the DOM for measurement
+            { attachTo: testsContainerElement },
+        ));
+    }
+
+    function resize(size: ISizeProps) {
+        wrapper.setProps(size);
+        return new Promise(resolve => setTimeout(resolve, 30));
```
Comment (giladgray) [inline packages/core/test/resize-sensor/resizeSensorTests.tsx]: sooo resizes need time to propagate (more than a frame, typically), so we must wrap each one in this brief timeout. fortunately, `async/await` make this rather nice to use instead of callback hell.

File: packages/core/src/components/resize-sensor/resizeSensor.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+export interface IResizeSensorProps {
+    /**
+     * Callback invoked when the wrapped element resizes.
+     */
+    onResize: (entries: IResizeEntry[]) => void;
+
+    /**
+     * If `true`, all parent DOM elements of the container will also be
+     * observed. If changes to a parent's size is detected, the overflow will be
+     * recalculated.
```
Comment (invliD) [inline packages/core/src/components/resize-sensor/resizeSensor.tsx]: the overflow part doesn't belong here

File: packages/core/src/components/resize-sensor/resizeSensor.tsx [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+     *
+     * Only enable this prop if the overflow should be recalculated when a
+     * parent element resizes in a way that does not also cause the
+     * `OverflowList` to resize.
```
Comment (invliD) [inline packages/core/src/components/resize-sensor/resizeSensor.tsx]: same here, update docs

File: packages/core/src/components/resize-sensor/resizeSensor.tsx [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+     * @default false
+     */
+    observeParents?: boolean;
+}
+
+export class ResizeSensor extends React.PureComponent<IResizeSensorProps> {
+    public static displayName = `${DISPLAYNAME_PREFIX}.Resize`;
```
Comment (invliD) [inline packages/core/src/components/resize-sensor/resizeSensor.tsx]: `ResizeSensor`

File: packages/core/test/resize-sensor/resizeSensorTests.tsx [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+import { IResizeSensorProps, ResizeSensor } from "../../src/components/resize-sensor/resizeSensor";
+
+describe("<ResizeSensor>", () => {
+    // this scope variable is assigned in mountResizeSensor() and used in resize()
+    let wrapper: ReactWrapper<IResizeTesterProps, any> | undefined;
+    const testsContainerElement = document.createElement("div");
+    document.documentElement.appendChild(testsContainerElement);
+
+    afterEach(() => {
+        // clean up list after each test, if it was used
```
Comment (invliD) [inline packages/core/test/resize-sensor/resizeSensorTests.tsx]: not a list anymore

File: packages/core/src/components/resize-sensor/resize-sensor.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
+---
+tag: new
+---
+
+@# Resize sensor
+
+`ResizeSensor` is a higher-order component that effectively provides a
```
Comment (themadcreator) [inline packages/core/src/components/resize-sensor/resize-sensor.md]: nit: this is a little wordy and im not sure it's even correct. Maybe rephrase:

    `ResizeSensor` is a component that provides a `"resize"` event for its single child.

But even then it's not "providing an event" in the sense of `.addEventListener`.

File: packages/core/test/resize-sensor/resizeSensorTests.tsx (hunk 3 above)
Comment (themadcreator) [reply packages/core/test/resize-sensor/resizeSensorTests.tsx]: oh.... this seems like a pretty big gotcha

Comment (themadcreator) [reply packages/core/test/resize-sensor/resizeSensorTests.tsx]: if this is going to be relevant to users of this component, you should add a comment to the `onResize` prop.
